include/clap/ext/draft/undo.h

```diff
@@ -112,35 +131,43 @@ typedef struct clap_host_undo {
    //
    // name: mandatory null terminated string describing the change, this is displayed to the user
    //
-   // deltas: optional, they are binary blobs used to perform the undo and redo. When not available
+   // delta: optional, it is a binary blobs used to perform the undo and redo. When not available
    // the host will save the plugin state and use state->load() to perform undo and redo.
+   // The plugin must be able to perform a redo operation using the delta, though the undo operation
+   // is only possible if delta_can_undo is true.
    //
    // Note: the provided delta may be used for incremental state saving and crash recovery. The
    // plugin can indicate a format version id and the validity lifetime for the binary blobs.
    // The host can use these to verify the compatibility before applying the delta.
    // If the plugin is unable to use a delta, a notification should be provided to the user and
-   // the crash recovery should perform a best effort job, at least restoring the latest saved state.
+   // the crash recovery should perform a best effort job, at least restoring the latest saved
+   // state.
    //
    // Special case: for objects with shared and synchronized state, changes shouldn't be reported
    // as the host already knows about it.
    // For example, plugin parameter changes shouldn't produce a call to change_made().
    //
+   // Note: if the plugin asked for this interface, then host_state->mark_dirty() will not create an
+   // implicit undo step.
+   //
+   // Note: if the plugin did load a preset or did something that leads to a large delta,
+   // it may consider not producing a delta (pass null) and let the host make a state snapshot
+   // instead.
+   //
+   // Note: if a plugin is producing a lot of changes within a small amount of time, the host
+   // may merge them into a single undo step.
+   //
    // [main-thread]
    void(CLAP_ABI *change_made)(const clap_host_t *host,
                                const char        *name,
-                               const void        *redo_delta,
-                               size_t             redo_delta_size,
-                               const void        *undo_delta,
-                               size_t             undo_delta_size);
+                               const void        *delta,
+                               size_t             delta_size,
+                               bool               delta_can_undo);
 
-   // Asks the host to perform the next undo step.
-   // This operation may be asynchronous.
+   // Asks the host to perform the next undo or redo step.
+   // This operation may be asynchronous and isn't available while the host is within a change.
```

Comment:
is it a misbehaviour if the plugin calls `undo`/`redo` if it isn't available?

There are 2 cases:
- the plugin did not subscribe to context updates -> it cannot know if its available -> should be always ok if the plugin calls `undo`/`redo`
  - question: will the host ignore the request or will it potentially delay the undo until a running action has finished? or is this up to the host?
- the plugin is subsribed to context updates -> it should know and it could be considered a misbehaviour if it calls undo/redo

So maybe change the conditions to `[main-thread & (!plugin-subscribed-to-undo-context | can-undo)]` / `[main-thread & (!plugin-subscribed-to-undo-context | can-redo)]`

Reply:
When the host is within a change, it simply set can_undo and can_redo to false.

Reply:
The plugin can use the host interface without implementing any of the plugin interfaces, so it is OK to call undo/redo without checking can_undo/redo. The host will simply ignore it in worst case.

Reply:
Maybe I should call it request_undo() and request_redo() instead? :thinking:

Reply:
To conclude this discussion, I'd not make it a misbehavior, because it'll make the interface more complicated to check and impl correctly.
Think that the plugins may be running in another process, with the gui on another thread, and all the state changes needs to be communicated, it is all asynchronous and it is maybe better to be a bit relaxed here.